File: flickr/src/main/java/com/pablocampos/flickrsample/model/ApiDataViewModel.java
package com.pablocampos.flickrsample.model;

import com.pablocampos.flickrsample.network.FlickrApi;
import com.pablocampos.flickrsample.network.FlickrInterface;

import java.util.ArrayList;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ApiDataViewModel extends ViewModel {


	private DataWrapper<FlickrFeed> dataWrapper;
	private MutableLiveData<DataWrapper<FlickrFeed>> liveData;



	public MutableLiveData<DataWrapper<FlickrFeed>> getLiveData () {
		if (liveData == null){
			dataWrapper = new DataWrapper<>();
			liveData = new MutableLiveData<>();
			loadFeeds("");
		}
		return liveData;
	}



	public void loadFeeds(final String tags){

		FlickrInterface service = FlickrApi.getClient().create(FlickrInterface.class);
		Call<ApiData> apiCall = service.getFlickrFeeds(tags);

		dataWrapper.setStatus(DataWrapper.Status.LOADING);
		liveData.postValue(dataWrapper);

		apiCall.enqueue(new Callback<ApiData>() {
			@Override
			public void onResponse(Call<ApiData> call, Response<ApiData> response) {

				// Stop refresh status and update data
				dataWrapper.setStatus(DataWrapper.Status.NONE);
				dataWrapper.setData(response.body().getItems());
				liveData.postValue(dataWrapper);
			}

			@Override
			public void onFailure(Call<ApiData> call, Throwable t) {

				// Stop refresh status
				dataWrapper.setStatus(DataWrapper.Status.ERROR);
				dataWrapper.setData(new ArrayList<FlickrFeed>());
				liveData.postValue(dataWrapper);
			}
		});
	}
}

File: flickr/src/main/kotlin/model/ApiDataViewModel.kt
package model

import android.content.Context
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import network.FlickrApi
import network.FlickrInterface
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class ApiDataViewModel : ViewModel() {


	private var dataWrapper: DataWrapper<FlickrFeed>? = null
	private var liveData: MutableLiveData<DataWrapper<FlickrFeed>>? = null


	fun getLiveData(context: Context): MutableLiveData<DataWrapper<FlickrFeed>> {
		if (liveData == null) {
			dataWrapper = DataWrapper()
			liveData = MutableLiveData()
			loadFeeds(context, "")
		}
		return liveData as MutableLiveData<DataWrapper<FlickrFeed>>
	}


	fun loadFeeds(context: Context, tags: String) {

		val service = FlickrApi.getClient(context).create(FlickrInterface::class.java)
		val apiCall = service.getFlickrFeeds(tags)

		dataWrapper!!.status = DataWrapper.Status.LOADING
		liveData!!.postValue(dataWrapper)

		apiCall.enqueue(object : Callback<ApiData> {
			override fun onResponse(call: Call<ApiData>, response: Response<ApiData>) {

				// Stop refresh status and update data
				dataWrapper!!.status = DataWrapper.Status.IDDLE
				dataWrapper!!.data = response.body()!!.items
				liveData!!.postValue(dataWrapper)
			}

			override fun onFailure(call: Call<ApiData>, t: Throwable) {

				// Stop refresh status
				dataWrapper!!.status = DataWrapper.Status.ERROR
				dataWrapper!!.data = ArrayList()
				liveData!!.postValue(dataWrapper)
			}
		})
	}
}
